client/galaxy/scripts/galaxy.interactive_environments.js

```diff
@@ -17,81 +17,159 @@ function display_spinner(){
         $('#main').append('<img id="spinner" src="' + galaxy_root + 'static/style/largespinner.gif" style="position:absolute;margin:auto;top:0;left:0;right:0;bottom:0;">');
 }
 
-function not_ready(timeout, timeout_values, timeout_time_max, timeout_time_step) {
-    if(timeout_values.count == 0){
-        display_spinner();
-        toastr.info(
-            "Galaxy is launching a container in which to run this interactive environment. Please wait...",
-            {'closeButton': true, 'tapToDismiss': false}
-        );
+
+/* Create a spin_state object used by spin() and spin_again() */
+function make_spin_state(type, ajax_timeout_init, ajax_timeout_max, ajax_timeout_step, sleep_init, sleep_max, sleep_step, log_attempts){
+    var s = {};
+    s.type = (typeof type !== 'undefined') ? type : "GIE spin";
+    s.ajax_timeout = (typeof ajax_timeout_init !== 'undefined') ? ajax_timeout_init : 2000;
```

Comment:
not a JS expert but I feel like i've usually seen this written as:

```javascript
var s = {
  ajax_timeout: ajax_timeout_init || 2000
  ...
}
```

Reply:
I agree with @erasche.
It could also be written as:
`s.ajax_timeout = ( ajax_timeout_init ) ? ajax_timeout_init : 2000`
These statements can check for `null`, `false`, `NaN` etc values as well unless there is a specific need to check only for "undefined".

Reply:
Thanks! I am about as far from a JS expert as you can get, so I appreciate the guidance. I'll fix it in the next commit.

Reply:
Just be careful when doing that.  Personally I prefer the explicit checks.  Using compound truth-test logic like this is what caused the missing history bug that took forever to find.

Reply:
Well, you make a good point @dannon. What with JS's weak typing and wacky implicit type conversions, maybe the explicit check here is more error-proof?

Reply:
It is definitely more error proof.  I do prefer one part of @erasche's suggestion, though.  I would write this more like:
```
var s = {
  ajax_timeout: (typeof ajax_timeout_init !== 'undefined') ? ajax_timeout_init : 2000;
  ...
}

Reply:
(that said, I'm totally ok with leaving this exactly as-is, and wouldn't rewrite it until after babel is available, which will allow for much better changes like default arguments to functions)

Reply:
I am also for explicit here, this has been identified as critical Galaxy problem so let's not leave any room for ambiguity.